src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,283 @@
+
+  for (size_t i = 0; i < jointTrajectory.joint_names.size(); ++i)
+  {
+    std::string joint_name = jointTrajectory.joint_names[i];
+    bool found_match = false;
+
+    for (size_t j = 0; j < space->getNumSubspaces(); ++j)
+    {
+      auto joint = space->getJointSpace(j)->getJoint();
+      if (joint_name == joint->getName())
+      {
+        // Check that the mapping is unique.
+        auto search = rosJointToMSSSJoint.find(i);
+        if (search != rosJointToMSSSJoint.end())
+        {
+          std::stringstream message;
+          message << "Both subspace[" << search->second << "] and ["
+            << j << "] map to joint " << i << " in jointTrajectory.";
+          throw std::invalid_argument{message.str()};
+        }
+        rosJointToMSSSJoint.emplace(std::make_pair(i, j));
+        found_match = true;
+        break;
+      }
+    }
+
+    // Check that there is a mapping for every joint.
+    if (!found_match)
+    {
+      std::stringstream message;
+      message << joint_name << " does not have a matching joint in space.";
+      throw std::invalid_argument{message.str()};
+    }
+  }
+
+  // Extract the first waypoint to infer the dimensionality of the trajectory.
+  Eigen::VectorXd currPosition, currVelocity, currAcceleration;
+  extractJointTrajectoryPoint(jointTrajectory, 0, numControlledJoints,
+    &currPosition, true, &currVelocity, false, &currAcceleration, false);
+
+  // To be used for reordering to match space's subspace order.
+  Eigen::VectorXd currPositionMSSS;
+  Eigen::MatrixXd segmentCoefficientsMSSS;
+
+  const auto& firstWaypoint = jointTrajectory.points.front();
+  auto currTimeFromStart = firstWaypoint.time_from_start.toSec();
+
+  const auto isPositionRequired = true;
+  const auto isVelocityRequired = (currVelocity.size() != 0);
+  const auto isAccelerationRequired = (currAcceleration.size() != 0);
+  if (isAccelerationRequired && !isVelocityRequired)
+  {
+    throw std::invalid_argument{
+      "Velocity is required since acceleration is specified."};
+  }
+
+  int numCoefficients;
+  if (isAccelerationRequired)
+    numCoefficients = 6; // quintic
+  else if (isVelocityRequired)
+    numCoefficients = 4; // cubic
+  else
+    numCoefficients = 2; // linear
+
+  // Convert the ROS trajectory message to an Aikido spline.
+  std::unique_ptr<SplineTrajectory> trajectory{new SplineTrajectory{space}};
+  auto currState = space->createState();
+
+  Eigen::MatrixXd conversionPermutation;
+  convertToPermutationMatrix(rosJointToMSSSJoint, &conversionPermutation);
+  const auto& waypoints = jointTrajectory.points;
+  for (size_t iwaypoint = 1; iwaypoint < waypoints.size(); ++iwaypoint)
+  {
+    Eigen::VectorXd nextPosition, nextVelocity, nextAcceleration;
+    extractJointTrajectoryPoint(jointTrajectory, iwaypoint, numControlledJoints,
+      &nextPosition, isPositionRequired,
+      &nextVelocity, isVelocityRequired,
+      &nextAcceleration, isAccelerationRequired);
```

Comment:
I'd like to perform the `reorder` operation here. These two operations are closely related because they convert from a ROS data structure to the Aikido data structure that we expect. The rest of this loop purely performs operations on Aikido data structures, e.g. fitting splines.